generic3g/FieldUtilities.F90

```diff
@@ -0,0 +1,131 @@
+#include "MAPL_Generic.h"
+
+module mapl3g_FieldUtilities
+   use MAPL_FieldPointerUtilities, only: FieldGetLocalElementCount
+   use mapl_ErrorHandling
+
+   use esmf
+   implicit none
+   private
+
+   public :: MAPL_FieldReallocate
+
+   interface MAPL_FieldReallocate
+      procedure :: reallocate
+   end interface MAPL_FieldReallocate
+
+   interface operator(==)
+      procedure :: ESMF_GeomEqual
+   end interface operator(==)
+
+   interface operator(/=)
+      procedure :: ESMF_GeomNotEqual
+   end interface operator(/=)
+
+contains
+
+
+   subroutine reallocate(field, geom, typekind, ungriddedUBound, rc)
+      type(ESMF_Field), intent(inout) :: field
+      type(ESMF_Geom), optional, intent(in) :: geom
+      type(ESMF_TypeKind_Flag), optional, intent(in) :: typekind
+      integer, optional, intent(in) :: ungriddedUBound(:)
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      type(ESMF_TypeKind_Flag) :: old_typekind, typekind_
+      type(ESMF_Geom) :: old_geom, geom_
+      logical :: skip_reallocate
+      integer :: ungriddedDimCount, rank
+      integer, allocatable :: localElementCount(:)
+      integer, allocatable :: old_ungriddedUBound(:)
+      integer, allocatable :: ungriddedUBound_(:), ungriddedLBound_(:)
+      integer :: i
+
+      skip_reallocate = .true.
+
+      call ESMF_FieldGet(field, typekind=old_typekind, geom=old_geom, ungriddedDimCount=ungriddedDimCount, rank=rank, _RC)
+      localElementCount = FieldGetLocalElementCount(field, _RC)
+      old_ungriddedUBound = localElementCount(rank-ungriddedDimCount+1:)
+
+      typekind_ = old_typekind
+      if (present(typekind)) typekind_ = typekind
+
+      geom_ = old_geom
+      if (present(geom)) geom_ = geom
+
+      ungriddedUBound_ = old_ungriddedUBound
+      if (present(ungriddedUBound)) ungriddedUBound_ = ungriddedUBound
+      _ASSERT(size(ungriddedUBound_) == size(old_ungriddedUBound), 'MAPL does not allow the rank of a field to change after creation.')
```

Comment:
Is it the _**total**_ number of dimensions of the field or the number of **_ungridded_** dimensions of the field you are checking?

Reply:
Just the ungridded dimensions.    We only allow the number of levels to change.    Changes in the gridded dimensions happen through a change in the geometry, which indirectly changes the extents of the gridded dimensions.   MAPL3 controls the ungridded explicitly.